File: src/main/java/dsl/TypedBranchBuilder.java
package dsl;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Predicate;

public class TypedBranchBuilder<S, T> {
    private final TypedTargetBuilder<S, T> parent;
    private final List<ConditionalBranch> branches = new ArrayList<>();

    TypedBranchBuilder(TypedTargetBuilder<S, T> parent) {
        this.parent = parent;
    }

    public TypedBranchConditionBuilder<S, T> when(Predicate<S> condition) {
        return new TypedBranchConditionBuilder<>(this, condition);
    }

    public TypedBranchActionBuilder<S, T> otherwise() {
        return new TypedBranchActionBuilder<>(this, obj -> true);
    }

    public TypedTargetBuilder<S, T> end() {
        parent.addRule(new BranchMapping(branches));
        return parent;
    }

    void addBranch(ConditionalBranch branch) {
        branches.add(branch);
    }
}

File: src/main/kotlin/xyz/mahmoudahmed/dsl/typed/TypedBranchBuilder.kt
package xyz.mahmoudahmed.dsl.typed

import xyz.mahmoudahmed.dsl.conditional.BranchMapping
import xyz.mahmoudahmed.dsl.conditional.ConditionalBranch

class TypedBranchBuilder<S, T>(
    private val parent: TypedTargetBuilder<S, T>
) {
    private val branches = mutableListOf<ConditionalBranch>()

    fun chooseIf(condition: (S) -> Boolean): TypedBranchConditionBuilder<S, T> {
        return TypedBranchConditionBuilder(this, condition)
    }

    fun otherwise(): TypedBranchActionBuilder<S, T> {
        return TypedBranchActionBuilder(this) { true }
    }

    fun end(): TypedTargetBuilder<S, T> {
        parent.addRule(BranchMapping(branches))
        return parent
    }

    internal fun addBranch(branch: ConditionalBranch) {
        branches.add(branch)
    }
}